File: src/clyde/book.py
from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import datetime


@dataclass(frozen=True)
class BookPlan:
    subject: str
    timestamp: datetime
    title_override: str | None = None

    @property
    def title(self) -> str:
        if self.title_override:
            return self.title_override
        return f"{self.timestamp.strftime('%Y-%m-%d %H%M')} - {self.subject}"

    @property
    def slug(self) -> str:
        if self.title_override:
            parsed = re.match(
                r"^(?P<date>\d{4})-(?P<month>\d{2})-(?P<day>\d{2}) "
                r"(?P<time>\d{4}) - (?P<subject>.+)$",
                self.title_override,
            )
            if parsed:
                stamp = (
                    f"{parsed.group('date')}{parsed.group('month')}{parsed.group('day')}-"
                    f"{parsed.group('time')}"
                )
                subject = _slug_text(parsed.group("subject"))
                return f"{stamp}-{subject}" if subject else stamp
            return _slug_text(self.title_override)
        stamp = self.timestamp.strftime("%Y%m%d-%H%M")
        subject = _slug_text(self.subject)
        return f"{stamp}-{subject}" if subject else stamp

    @property
    def source_prefix(self) -> str:
        return f"{self.title} :: "

    @classmethod
    def create(cls, subject: str, *, now: datetime | None = None) -> "BookPlan":
        cleaned = " ".join(subject.split())
        if not cleaned:
            raise ValueError("book subject must not be empty")
        timestamp = now or datetime.now().astimezone()
        return cls(subject=cleaned, timestamp=timestamp)

    @classmethod
    def from_title(cls, title: str) -> "BookPlan":
        cleaned = " ".join(title.split())
        if not cleaned:
            raise ValueError("book title must not be empty")
        return cls(subject=cleaned, timestamp=datetime.now().astimezone(), title_override=cleaned)


def _slug_text(value: str) -> str:
    return re.sub(r"[^a-z0-9]+", "-", value.lower()).strip("-")

File: internal/clyde/book.go
package clyde

import (
	"regexp"
	"strings"
	"time"
)

type BookPlan struct {
	Subject       string
	Timestamp     time.Time
	TitleOverride string
}

func NewBookPlan(subject string, now time.Time) (BookPlan, error) {
	cleaned := strings.Join(strings.Fields(subject), " ")
	if cleaned == "" {
		return BookPlan{}, errf("book subject must not be empty")
	}
	if now.IsZero() {
		now = time.Now()
	}
	return BookPlan{Subject: cleaned, Timestamp: now}, nil
}

func BookPlanFromTitle(title string) (BookPlan, error) {
	cleaned := strings.Join(strings.Fields(title), " ")
	if cleaned == "" {
		return BookPlan{}, errf("book title must not be empty")
	}
	return BookPlan{Subject: cleaned, Timestamp: time.Now(), TitleOverride: cleaned}, nil
}

func (p BookPlan) Title() string {
	if p.TitleOverride != "" {
		return p.TitleOverride
	}
	return p.Timestamp.Format("2006-01-02 1504") + " - " + p.Subject
}

func (p BookPlan) Slug() string {
	if p.TitleOverride != "" {
		re := regexp.MustCompile(`^(\d{4})-(\d{2})-(\d{2}) (\d{4}) - (.+)$`)
		match := re.FindStringSubmatch(p.TitleOverride)
		if match != nil {
			stamp := match[1] + match[2] + match[3] + "-" + match[4]
			subject := slugText(match[5])
			if subject == "" {
				return stamp
			}
			return stamp + "-" + subject
		}
		return slugText(p.TitleOverride)
	}
	stamp := p.Timestamp.Format("20060102-1504")
	subject := slugText(p.Subject)
	if subject == "" {
		return stamp
	}
	return stamp + "-" + subject
}

func (p BookPlan) SourcePrefix() string {
	return p.Title() + " :: "
}

func slugText(value string) string {
	re := regexp.MustCompile(`[^a-z0-9]+`)
	return strings.Trim(re.ReplaceAllString(strings.ToLower(value), "-"), "-")
}
